### 06. Gradients

Let's backprop now. Because `out` contains a single scalar, `out.backward()` is equivalent to `out.backward(torch.tensor(1.))`.

In [1]:
import torch

In [2]:
x = torch.ones(2, 2, requires_grad=True)
y = x + 2
z = y * y *3
out = z.mean()
print(out)

tensor(27., grad_fn=<MeanBackward0>)


print gradients d(out)/dx

In [3]:
out.backward()
print(x.grad)

tensor([[4.5000, 4.5000],
        [4.5000, 4.5000]])


let's call the out Tensor as $o$. We have that $o=\frac{1}{4}\sum\limits_{i} z_{i}$ and $z_{i}=3(x_{i}+2)^2$ and $z_{i}|_{x_{i}=1}=27$

Therefore, $\frac{\partial o}{\partial x_{i}}=\frac{3}{2}(x_{i}+2)$, hence $\frac{\partial o}{\partial x_{i}}|_{x_{i}=1}=\frac{3}{2}(1+2)=\frac{9}{2}=4.5$